pkg/controller/podautoscaler/metrics/rest_metrics_client_test.go

```diff
@@ -0,0 +1,274 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+	core "k8s.io/client-go/testing"
+	kv1 "k8s.io/kubernetes/pkg/api/v1"
+	autoscalingapi "k8s.io/kubernetes/pkg/apis/autoscaling/v2alpha1"
+	metricsfake "k8s.io/metrics/pkg/client/clientset_generated/clientset/fake"
+	cmfake "k8s.io/metrics/pkg/client/custom_metrics/fake"
+
+	cmapi "k8s.io/metrics/pkg/apis/custom_metrics/v1alpha1"
+	metricsapi "k8s.io/metrics/pkg/apis/metrics/v1alpha1"
+
+	"github.com/stretchr/testify/assert"
+
+	// we need the API types for rest mapping lookup
+	_ "k8s.io/client-go/pkg/api/install"
+	_ "k8s.io/client-go/pkg/apis/extensions/install"
+)
+
+type restClientTestCase struct {
+	desiredMetricValues PodMetricsInfo
+	desiredError        error
+
+	targetTimestamp      int
+	reportedMetricPoints []metricPoint
+	reportedPodMetrics   [][]int64
+	singleObject         *autoscalingapi.CrossVersionObjectReference
+
+	namespace    string
+	selector     labels.Selector
+	resourceName v1.ResourceName
+	metricName   string
+}
+
+func (tc *restClientTestCase) prepareTestClient(t *testing.T) (*metricsfake.Clientset, *cmfake.FakeCustomMetricsClient) {
+	namespace := "test-namespace"
+	tc.namespace = namespace
+	podNamePrefix := "test-pod"
+	podLabels := map[string]string{"name": podNamePrefix}
+	tc.selector = labels.SelectorFromSet(podLabels)
+
+	// it's a resource test if we have a resource name
+	isResource := len(tc.resourceName) > 0
+
+	fakeMetricsClient := &metricsfake.Clientset{}
+	fakeCMClient := &cmfake.FakeCustomMetricsClient{}
+
+	if isResource {
+		fakeMetricsClient.AddReactor("list", "podmetricses", func(action core.Action) (handled bool, ret runtime.Object, err error) {
+			metrics := &metricsapi.PodMetricsList{}
+			for i, containers := range tc.reportedPodMetrics {
+				metric := metricsapi.PodMetrics{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fmt.Sprintf("%s-%d", podNamePrefix, i),
+						Namespace: namespace,
+						Labels:    podLabels,
+					},
+					Timestamp:  metav1.Time{Time: fixedTimestamp.Add(time.Duration(tc.targetTimestamp) * time.Minute)},
```

Comment:
I don't understand, why Duration is used here. It seems to be redundant and wrong.